File: app/src/main/java/com/health/openworkout/gui/preference/TimePreferenceDialog.java
package com.health.openworkout.gui.preference;

import android.os.Bundle;
import android.text.format.DateFormat;
import android.view.View;
import android.widget.TimePicker;

import androidx.preference.DialogPreference;
import androidx.preference.PreferenceDialogFragmentCompat;

import com.health.openworkout.R;

import java.util.Calendar;

public class TimePreferenceDialog extends PreferenceDialogFragmentCompat {
    private Calendar calendar;
    private TimePicker timePicker;

    public static TimePreferenceDialog newInstance(String key) {
        final TimePreferenceDialog fragment = new TimePreferenceDialog();
        final Bundle b = new Bundle(1);
        b.putString(ARG_KEY, key);
        fragment.setArguments(b);

        return fragment;
    }

    @Override
    protected void onBindDialogView(View view) {
        super.onBindDialogView(view);

        timePicker = view.findViewById(R.id.timePicker);
        calendar = Calendar.getInstance();

        Long timeInMillis = null;
        DialogPreference preference = getPreference();

        if (preference instanceof TimePreference) {
            TimePreference timePreference = (TimePreference) preference;
            timeInMillis = timePreference.getTimeInMillis();
        }

        if (timeInMillis != null) {
            calendar.setTimeInMillis(timeInMillis);
            boolean is24hour = DateFormat.is24HourFormat(getContext());

            timePicker.setIs24HourView(is24hour);
            timePicker.setHour(calendar.get(Calendar.HOUR_OF_DAY));
            timePicker.setMinute(calendar.get(Calendar.MINUTE));
        }
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
        if (positiveResult) {
            int hours;
            int minutes;

            hours = timePicker.getHour();
            minutes = timePicker.getMinute();

            calendar.set(Calendar.HOUR_OF_DAY, hours);
            calendar.set(Calendar.MINUTE, minutes);

            long timeInMillis = calendar.getTimeInMillis();

            DialogPreference preference = getPreference();
            if (preference instanceof TimePreference) {
                TimePreference timePreference = ((TimePreference) preference);
                if (timePreference.callChangeListener(timeInMillis)) {
                    timePreference.setTimeInMillis(timeInMillis);
                    timePreference.setSummary(DateFormat.getTimeFormat(getContext()).format(calendar.getTime()));
                }
            }
        }
    }
}

File: app/src/main/java/com/health/openworkout/gui/preference/TimePreferenceDialog.kt
package com.health.openworkout.gui.preference

import android.app.Dialog
import android.os.Bundle
import android.text.format.DateFormat

import androidx.preference.PreferenceDialogFragmentCompat

import com.google.android.material.timepicker.MaterialTimePicker
import com.google.android.material.timepicker.TimeFormat
import com.health.openworkout.R

import java.util.Calendar

class TimePreferenceDialog : PreferenceDialogFragmentCompat() {
    private lateinit var calendar: Calendar
    private lateinit var timePicker: MaterialTimePicker

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        calendar = Calendar.getInstance()

        (preference as? TimePreference)?.timeInMillis?.let { timeInMillis ->
            calendar.timeInMillis = timeInMillis
            val is24hour = DateFormat.is24HourFormat(context)
            val timeFormat = if (is24hour) TimeFormat.CLOCK_24H else TimeFormat.CLOCK_12H

            timePicker = MaterialTimePicker.Builder()
                .setTimeFormat(timeFormat)
                .setHour(calendar.get(Calendar.HOUR_OF_DAY))
                .setMinute(calendar.get(Calendar.MINUTE))
                .setTitleText(R.string.select_time)
                .build()

            timePicker.addOnPositiveButtonClickListener {
                val hours = timePicker.hour
                val minutes = timePicker.minute

                calendar.apply {
                    set(Calendar.HOUR_OF_DAY, hours)
                    set(Calendar.MINUTE, minutes)
                }

                val timeInMillis = calendar.timeInMillis

                (preference as? TimePreference)?.let { pref ->
                    if (pref.callChangeListener(timeInMillis)) {
                        pref.timeInMillis = timeInMillis
                        pref.summary = DateFormat.getTimeFormat(context).format(calendar.time)
                    }
                }
            }

            timePicker.show(parentFragmentManager, "MATERIAL_TIME_PICKER")
        }

        return super.onCreateDialog(savedInstanceState)
    }

    override fun onDialogClosed(positiveResult: Boolean) {}

    companion object {
        @JvmStatic
        fun newInstance(key: String?): TimePreferenceDialog {
            return TimePreferenceDialog().apply {
                arguments = Bundle(1).apply {
                    putString(ARG_KEY, key)
                }
            }
        }
    }
}
